src/lib/InstantSearch.ts

```diff
@@ -633,32 +639,27 @@ See ${createDocumentationLink({
         ? uiState(this.mainIndex.getWidgetUiState({}) as TUiState)
         : uiState;
 
-    const setIndexHelperState = (indexWidget: IndexWidget) => {
-      const nextIndexUiState = nextUiState[indexWidget.getIndexId()] || {};
-
-      if (__DEV__) {
-        checkIndexUiState({
-          index: indexWidget,
-          indexUiState: nextIndexUiState,
-        });
-      }
-
-      indexWidget.getHelper()!.setState(
-        indexWidget.getWidgetSearchParameters(indexWidget.getHelper()!.state, {
-          uiState: nextIndexUiState,
-        })
-      );
-
-      indexWidget
-        .getWidgets()
-        .filter(isIndexWidget)
-        .forEach(setIndexHelperState);
-    };
-
-    setIndexHelperState(this.mainIndex);
+    if (this.onStateChange && callOnStateChange) {
+      this.onStateChange({
+        uiState: nextUiState,
+        setUiState: (finalUiState) => {
+          setIndexHelperState(
+            typeof finalUiState === 'function'
+              ? finalUiState(nextUiState)
+              : finalUiState,
+            this.mainIndex
+          );
+
+          this.scheduleSearch();
+          this.onInternalStateChange();
+        },
+      });
+    } else {
+      setIndexHelperState(nextUiState, this.mainIndex);
```

Comment:
this could be in the same branch if we had a default to onStateChange (same with the if inside helper.search override in index), should we do that in a major version, now or never?